R/MSnExperiment-class.R

```diff
@@ -783,27 +1383,126 @@ setMethod("filterFile", "MSnExperiment", function(object, file) {
+                           paste0("Filter: select spectra with polarity  ",
+                                  paste0(polarity., collapse = " "),
+                                  " [", date(), "]"))
+    object
+})
+
+#' @rdname MSnExperiment
+setMethod("filterRt", "MSnExperiment", function(object, rt, msLevel.) {
+    if (missing(rt)) return(object)
+    if (!(length(rt) == 2 & is.numeric(rt)))
+        stop("'rt' must be a numeric of length 2", call. = FALSE)
+    if (missing(msLevel.))
+        msLevel. <- unique(msLevel(object))
+    else if (!is.numeric(msLevel.))
+        stop("'msLevel' must be numeric", call. = FALSE)
+    rt <- range(rt)
+    msLevel. <- unique(msLevel.)
+    sel_ms <- msLevel(object) %in% msLevel.
+    sel_rt <- rtime(object) >= rt[1] & rtime(object) <= rt[2] & sel_ms
+    object <- object[sel_rt | !sel_ms, drop = FALSE]
+    object@processing <- c(object@processing,
+                           paste0("Filter: select retention time [",
+                                  rt[1], "..", rt[2], "] on MS level(s) ",
+                                  paste0(unique(msLevel.), collapse = " "),
+                                  " [", date(), "]"))
+    object
+})
+
+#' @rdname MSnExperiment
+setMethod("splitByFile", c("MSnExperiment", "factor"), function(object, f) {
+    if (length(f) != length(fileNames(object)))
+        stop("length of 'f' has to match the length of samples/files in 'object'.")
+    idxs <- lapply(levels(f), function(z) which(f == z))
+    res <- lapply(idxs, function(z) {
+        filterFile(object, file = z)
+    })
+    names(res) <- levels(f)
+    res
+})
+
 ##============================================================
 ##  --  DATA MANIPULATION METHODS
 ##
 ##------------------------------------------------------------
 
 #' @rdname MSnExperiment
-setMethod("removePeaks", "MSnExperiment", function(object, t = "min",
-                                                   verbose = isMSnbaseVerbose(),
-                                                   msLevel.) {
-    if (!is.numeric(t) & t != "min")
-        stop("Argument 't' has to be either numeric of 'min'.")
+setMethod("bin", "MSnExperiment", function(object, binSize = 1L, msLevel.) {
     if (missing(msLevel.))
-        msLevel. <- base::sort(unique(object@spectraData$msLevel))
-    if (!is.numeric(msLevel.))
-        stop("'msLevel' must be numeric.")
-    object <- addProcessingStep(object, "removePeaks", t = t,
+        msLevel. <- base::sort(unique(msLevel(object)))
+    else if (!is.numeric(msLevel.))
+        stop("'msLevel' must be numeric", call. = FALSE)
+    if (!any(unique(msLevel(object)) %in% msLevel.)) {
+        warning("No spectra of the specified MS level present.")
+        return(object)
+    }
+    mzr <- range(unlist(spectrapply(filterMsLevel(object, msLevel. = msLevel.),
+                                    FUN = function(z) {
+                                        range(z@mz, na.rm = TRUE)
```

Comment:
We know that `@mz` is ordered, so we could use the first and last mz instead of `range`:
```suggestion
                                        z@mz[c(1L, length(z@mz)]
```
(this will cause problems if `@mz == double(0)` (returns `NA`) but `range` would return (`-Inf`, `Inf`) in that case)

```r
library("microbenchmark")
sf <- dir(system.file("sciex", package = "msdata"), full.names = TRUE)
sciex <- readMSnExperiment(sf, backend = BackendMemory())

mzr1 <- function(object) {
    range(unlist(spectrapply(object, FUN = function(z)range(z@mz, na.rm = TRUE))))
}
mzr2 <- function(object) {
    range(unlist(spectrapply(object, FUN = function(z)z@mz[c(1L, length(z@mz))])))
}

all.equal(mzr1(sciex), mzr2(sciex))
# [1] TRUE
microbenchmark(mzr1(sciex), mzr2(sciex), times=2)
# Unit: seconds
#         expr      min       lq     mean   median       uq      max neval
#  mzr1(sciex) 2.085682 2.085682 2.086945 2.086945 2.088208 2.088208     2
#  mzr2(sciex) 1.766330 1.766330 1.792665 1.792665 1.819000 1.819000     2
```

Reply:
OK. We just have to use then `na.rm = TRUE` in the `range` call.